sqlglot/dialects/dialect.py

```diff
@@ -42,6 +42,20 @@ class Dialects(str, Enum):
 class _Dialect(type):
     classes: t.Dict[str, t.Type[Dialect]] = {}
 
+    def __eq__(cls, other: DialectType) -> bool:  # type: ignore
+        if cls is other:
+            return True
+        if isinstance(other, str):
+            return cls is cls.get(other.lower())
+        if isinstance(other, Dialect):
+            return cls is type(other)
+
+        return False
+
+    def __hash__(cls) -> int:
+        # This will ensure that equal _Dialects have the same hashes
+        return super().__hash__()
```

Comment:
can you explain again why this was necessary

Reply:
Without overriding it I was getting the following error:

```
TypeError: unhashable type: '_Dialect'
```

<details close>
<summary>Full stack trace</summary>

```python
======================================================================
ERROR: tests.test_transpile (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: tests.test_transpile
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "sqlglot/tests/test_transpile.py", line 5, in <module>
    from sqlglot import parse_one, transpile
  File "sqlglot/sqlglot/__init__.py", line 13, in <module>
    from sqlglot.dialects.dialect import Dialect as Dialect, Dialects as Dialects
  File "sqlglot/sqlglot/dialects/__init__.py", line 59, in <module>
    from sqlglot.dialects.bigquery import BigQuery
  File "sqlglot/sqlglot/dialects/bigquery.py", line 9, in <module>
    from sqlglot.dialects.dialect import (
  File "sqlglot/sqlglot/dialects/dialect.py", line 228, in <module>
    DialectType = t.Union[str, Dialect, t.Type[Dialect], None]
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 312, in inner
    return func(*args, **kwds)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 403, in __getitem__
    return self._getitem(self, parameters)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 516, in Union
    parameters = _remove_dups_flatten(parameters)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 281, in _remove_dups_flatten
    return tuple(_deduplicate(params))
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 255, in _deduplicate
    all_params = set(params)
TypeError: unhashable type: '_Dialect'
```

Seems like the error is caused because of the `DialectType = Union[... t.Type[Dialect] ...]` definition; Python internally stores these types in a set which is probably a hash under the hood, and so when it tries to store `t.Type[Dialect]` it fails with the above error.

</details>

See https://docs.python.org/3/reference/datamodel.html#object.__hash__

Reply:
Updated so that we do `__hash__ = type.__hash__` as per the documentation.

Reply:
Updated yet again so that we actually hash the dialect's _name_ now, instead of using `type.__hash__`. This was needed to enable checks like `dialect in {"snowflake", "bigquery"}`.